FFmpegInterop/SampleProvider.cpp

```diff
@@ -0,0 +1,415 @@
+		// If there's no packets in the queue this stream is now EOS.
+		if (!m_isEOS && m_packetQueue.empty())
+		{
+			m_isEOS = true;
+
+			if (m_isSelected)
+			{
+				TraceLoggingWrite(g_FFmpegInteropProvider, "EndOfStream", TraceLoggingLevel(TRACE_LEVEL_VERBOSE), TraceLoggingPointer(this, "this"),
+					TraceLoggingValue(m_stream->index, "StreamId"));
+			}
+		}
+	}
+
+	void SampleProvider::Flush() noexcept
+	{
+		m_packetQueue.clear();
+		m_isDiscontinuous = true;
+	}
+
+	void SampleProvider::QueuePacket(_In_ AVPacket_ptr packet)
+	{
+		if (m_isSelected)
+		{
+			m_packetQueue.push_back(move(packet));
+		}
+	}
+
+	void SampleProvider::GetSample(_Inout_ const MediaStreamSourceSampleRequest& request)
+	{
+		TraceLoggingWrite(g_FFmpegInteropProvider, "SampleRequested", TraceLoggingLevel(TRACE_LEVEL_VERBOSE), TraceLoggingPointer(this, "this"),
+			TraceLoggingValue(m_stream->index, "StreamId"));
+
+		// Make sure this stream is selected
+		THROW_HR_IF(MF_E_INVALIDREQUEST, !m_isSelected);
+
+		// Get the sample data, timestamp, duration, and properties
+		auto [buf, pts, dur, properties] = GetSampleData();
+
+		// Make sure the PTS is set
+		if (pts == AV_NOPTS_VALUE)
+		{
+			pts = m_nextSamplePts;
+		}
+
+		// Calculate the PTS for the next sample
+		m_nextSamplePts = pts + dur;
+
+		// Convert time base from FFmpeg to MF
+		pts = ConvertFromAVTime(pts - m_startOffset, m_stream->time_base, HNS_PER_SEC);
+		dur = ConvertFromAVTime(dur, m_stream->time_base, HNS_PER_SEC);
+
+		// Create the sample
+		MediaStreamSample sample{ MediaStreamSample::CreateFromBuffer(buf, static_cast<TimeSpan>(pts)) };
+		sample.Duration(TimeSpan{ dur });
+		sample.Discontinuous(m_isDiscontinuous);
+
+		MediaStreamSamplePropertySet extendedProperties{ sample.ExtendedProperties() };
+		for (const auto& [key, value] : properties)
+		{
+			extendedProperties.Insert(key, value);
+		}
+
+		m_isDiscontinuous = false;
+
+		request.Sample(sample);
+
+		TraceLoggingWrite(g_FFmpegInteropProvider, "SampleRequestFilled", TraceLoggingLevel(TRACE_LEVEL_VERBOSE), TraceLoggingPointer(this, "this"),
+			TraceLoggingValue(m_stream->index, "StreamId"),
+			TraceLoggingValue(sample.Timestamp().count(), "TimestampHNS"),
+			TraceLoggingValue(sample.Duration().count(), "DurHNS"));
+	}
+
+	tuple<IBuffer, int64_t, int64_t, map<GUID, IInspectable>> SampleProvider::GetSampleData()
+	{
+		AVPacket_ptr packet{ GetPacket() };
+
+		const int64_t pts{ packet->pts };
+		const int64_t dur{ packet->duration };
+
+		// Set sample properties
+		map<GUID, IInspectable> properties;
+
+		if ((packet->flags & AV_PKT_FLAG_KEY) != 0)
+		{
+			properties[MFSampleExtension_CleanPoint] = PropertyValue::CreateUInt32(true);
+		}
+
+		return { make<FFmpegInteropBuffer>(move(packet)), pts, dur, move(properties) };
+	}
+
+	AVPacket_ptr SampleProvider::GetPacket()
+	{
+		// Continue reading until there is an appropriate packet in the stream
+		while (m_packetQueue.empty())
+		{
+			m_reader.ReadPacket();

```

Comment:
>m_reader [](start = 3, length = 8)

nit: I know you just followed the pre-existing design here, but IMO this approach of plumbing the reader into each sample provider and having the sample providers call ReadPacket() isn't very clean. Specifically because calling ReadPacket() on the reader from a sample provider would also impact the state (queued packets) of other sample providers. It would be cleaner if we had this logic within the parent FFmpegInteropMSS class instead (i.e in FFmpegInteropMSS::OnSampleRequested, we could check whether the sample provider has queued packets, and if not, perform packet reads until we've queued a packet for it). We can probably leave this as it is for now, but it's something to consider as a future cleanup task.

Reply:
Another issue which would be alleviated by having this logic in the parent FFmpegInteropMSS class is the fact that there is an implicit contract here that m_reader.ReadPacket will eventually either fail (throw) or cause a packet to be added to this sample provider's packet queue. This is kind of fragile - there is already a case where the sample provider won't add a packet to its queue when the reader calls QueuePacket() on it (when the stream is not selected). We currently have a mutex in the parent FFmpegInteropMSS class to prevent DeselectStream from being called on a sample provider while this loop is executing - it would be easier to reason about this and prevent future regressions in this area if this loop was also in FFmpegInteropMSS.

Reply:
Yeah we can revisit this in the future. One complication is that the FFmpegInteropMSS doesn't know ahead of time how many packets a sample provider will need to produce a sample.  The current design nicely encapsulates special case logic in the sample provider.

Examples:
- Uncompressed sample providers need a varying number of packets to produce a sample. At EOF when the decoder enters draining mode the sample provider may still be able to generate samples without any additional packet input.
- The subtitle sample provider defers sample requests when its packet queue is empty.